File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& arr) {
        // 1<<n => 2^n
        // for subsets total subsets will be 2^n
        int n=arr.size();
        vector<vector<int>>ans;

        int subset=1<<n;
        for(int i=0;i<subset;i++){
            vector<int>list;
            for(int j=0;j<n;j++){
                if(i & (1<<j)){
                    list.push_back(arr[j]);
                }
            }
            ans.push_back(list);
        }
        return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {

    void func(int[] arr,List<Integer>temp,int i,List<List<Integer>>ans){

        if(i==arr.length){
            ans.add(new ArrayList<>(temp));
            return;
        }
        temp.add(arr[i]);
        func(arr,temp,i+1,ans);

        temp.remove(temp.size() - 1);
        func(arr,temp,i+1,ans);
    }

    public List<List<Integer>> subsets(int[] arr) {
        
        List<List<Integer>>ans=new ArrayList<>();
        List<Integer> temp = new ArrayList<>();

        func(arr,temp,0,ans);

        return ans;
    }
}